File: Day10-19/13.py
def parse_line(l):
    fields = l.split()
    depth = int(fields[0][:-1])
    rnge = int(fields[1])
    return (depth, rnge)


def parse_ranges(lines):
    max_depth, _ = parse_line(lines[-1])
    out = [0 for i in range(max_depth+1)]
    for l in lines:
        fields = l.split()
        depth = int(fields[0][:-1])
        rnge = int(fields[1])
        out[depth] = rnge
    return out


def trip_severity(ranges):
    severity = 0
    for time, rnge in enumerate(ranges):
        if rnge != 0 and time % (2*rnge-2) == 0:
            severity += time*rnge
    return severity


def is_caught(ranges, delay):
    for time, rnge in enumerate(ranges):
        if rnge != 0 and (time+delay) % (2*rnge-2) == 0:
            return True
    return False


def min_delay(ranges):
    delay = 0
    while is_caught(ranges, delay):
        delay += 1
    return delay

if __name__ == '__main__':
    with open('13.txt') as f:
        lines = f.readlines()
    ranges = parse_ranges(lines)
    part1 = trip_severity(ranges)
    print("Part 1: {}".format(part1))

    part2 = min_delay(ranges)
    print("Part 2: {}".format(part2))

File: Day10-19/13.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

type Segment struct {
	rng   int
	depth int
}

func parseSegment(line string) Segment {
	fields := strings.Fields(line)
	depth, _ := strconv.Atoi(fields[0][:len(fields[0])-1])
	rng, _ := strconv.Atoi(fields[1])
	return Segment{rng: rng, depth: depth}
}

func parseLines(lines []string) []Segment {
	segments := make([]Segment, 0)
	for _, line := range lines {
		segments = append(segments, parseSegment(line))
	}
	return segments
}

func tripSeverity(segments []Segment) int {
	severity := 0
	for _, segment := range segments {
		if (2*segment.rng-2) == 0 || segment.depth%(2*segment.rng-2) == 0 {
			severity += segment.rng * segment.depth
		}
	}
	return severity
}

func getsCaught(segments []Segment, delay int) bool {
	for _, segment := range segments {
		if (delay+segment.depth)%(2*segment.rng-2) == 0 {
			return true
		}
	}
	return false
}

func minDelay(segments []Segment) int {
	delay := 0
	for delay = 0; getsCaught(segments, delay); delay++ {
	}
	return delay
}

func main() {
	dat, _ := ioutil.ReadFile("./13.txt")
	lines := strings.Split(string(dat), "\n")
	segments := parseLines(lines)

	part1 := tripSeverity(segments)
	fmt.Printf("Part 1: %d\n", part1)

	part2 := minDelay(segments)
	fmt.Printf("Part 2: %d\n", part2)
}
